### ISSUE TITLE
kc_locale parameter not working as expected in keycloak 25 (keycloak.v3 account theme)

### ISSUE THREAD

**Issue by AkshadaMulay** [keycloak/keycloak#36116]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

account/ui

### Describe the bug

- We migrated our application from Keycloak 16 to Keycloak 25. 
- In Keycloak 16, when we switched the application's user interface language (example: to Greek), and clicked on the profile button, the Keycloak dashboard used to open in the selected language. This behavior worked because the kc_locale parameter in the URL correctly mapped to the corresponding locale and rendered the Keycloak dashboard in the selected language.

1. After upgrading to Keycloak 25:

- The kc_locale parameter does not work as expected.
- Regardless of the value passed to kc_locale in the URL (Example: account/?referrer=client_id&kc_locale=el OR account/?referrer=client_id&kc_locale=el-GR ), the Keycloak dashboard defaults to English.
- Even though we have updated the Keycloak theme to include all the supported locales and their variations, the issue persists.

However:

- When we navigate to the dashboard, the 'Select a locale' dropdown in the profile allows us to manually select a language. Once selected, the dashboard reflects the chosen locale (example: Greek). Reference image shown below:
- 
![image](https://github.com/user-attachments/assets/0b8ee816-a2df-4d29-a735-fcf9accfe041)

- The issue is isolated to the kc_locale parameter not being respected when passed via the URL.

### Version

Keycloak 25

### Regression

- [X] The issue is a regression

### Expected behavior

When the kc_locale parameter is passed in the URL, Keycloak should use the value of this parameter to determine the language of the user interface. For example:

If kc_locale=el, the dashboard should open in Greek.
If kc_locale=es-CO, the dashboard should open in Spanish (Colombian)

### Actual behavior

The kc_locale parameter is ignored, and the Keycloak dashboard always defaults to English, regardless of the parameter's value.

### How to Reproduce?

Update Keycloak to version 25.

- Configure the Keycloak realm to enable localization and add supported locales (e.g., en, el-GR, el-CY, es-CO, fr-CH, etc.).
- Set up a custom Keycloak theme with all supported messages_<locale>.properties files.
- Access the Keycloak profile URL with the kc_locale parameter, e.g.:
- 
![image](https://github.com/user-attachments/assets/273f3ecf-9a40-46ea-87f8-2c7c0137746a)

- Observe that the language defaults to English, regardless of the kc_locale parameter's value.
- Manually select a language from the "Select a locale" dropdown in the dashboard. 
- Observe that the locale changes successfully when selected manually.

### Anything else?

1. Is the kc_locale parameter still supported in Keycloak 25? If so, has its behavior or usage been modified in any way compared to earlier versions? Documentation on this change would be helpful.
2. This behavior was verified in a clean Keycloak 25 setup with localization and internationalization properly configured in the realm.json file:
"internationalizationEnabled": true,
    "supportedLocales": [
        "pt-BR",
        "en",
        "fr",
        "fr-CH",
        "fr-CA",
        "fr-BE",
        "fr-FR",
        "es",
        "es-ES",
        "es-BR",
        "es-CO",
        "es-MX",
        "es-DO",
        "es-CR",
        "es-PA",
        "es-CL",
        "el-GR",
        "el-CY",
        "pt-PT",
    ],
    "defaultLocale": "en"

3. Additionally, the custom theme includes theme.properties with:
locales=el-GR,el-CY,en,es,es-ES,es-BR,es-CO,es-MX,es-DO,es-CR,es-PA,es-CL,fr,fr-CH,fr-CA,fr-BE,fr-FR,pt-BR,pt-PT
And corresponding messages_<locale>.properties files.

**Comment by AkshadaMulay** [comment]
I have tested this issue on the latest nightly release of Keycloak (KEYCLOAK_VERSION=999.0.0-SNAPSHOT), and the issue is still reproducible. The kc_locale parameter is not working as expected in the following scenario:
1. Configure a clean instance of the latest Keycloak nightly build.
2. Enable Localization in the realm settings, add supported locales (e.g., en, el, es-CO), and set a default locale.
3. Apply a custom theme with the correct messages_<locale>.properties files and theme.properties configured with supported locales mentioned in the above comment
4. Navigate to the Keycloak profile URL with the kc_locale parameter, e.g.:
![image](https://github.com/user-attachments/assets/c8c3c331-2b5c-4091-886c-e3c0cffc08d2)
5. Observe that the dashboard language defaults to English instead of the language specified in the kc_locale parameter
![image](https://github.com/user-attachments/assets/f5664fd9-fcfc-4e16-a226-a17948891d32)

- Current Behavior:

The kc_locale parameter is ignored, and the dashboard always defaults to English.

- Expected Behavior:

The kc_locale parameter should determine the language of the user interface. For example:
kc_locale=el should display the dashboard in Greek.
kc_locale=es-CO should display the dashboard in Spanish (Colombian)


- Additional Notes
The "Select a locale" dropdown in the profile allows the language to be changed manually, and the language is reflected correctly in the UI, confirming that the supported locales are properly configured.

- **This behavior was tested on both Keycloak 25 and the latest nightly build and persists in both versions.**

- Could you confirm whether the kc_locale parameter is still supported in the latest releases? If so, has its behavior changed, or does this indicate a potential regression?

**Comment by AkshadaMulay** [comment]
Hello! I wanted to kindly follow up on this issue to check if there are any updates or clarifications regarding the behavior of the kc_locale parameter in the latest release.
As mentioned earlier, the issue persists in both Keycloak 25 and the nightly build (tested on KEYCLOAK_VERSION=999.0.0-SNAPSHOT), and we are hoping to confirm whether this behavior is an intentional change or a potential regression. 

Thank you for your time and support!

**Comment by AkshadaMulay** [comment]
Hello!

I noticed that the action/old-release tag has been applied to this issue. However, I have tested the behavior in the nightly build and the issue persists there as well.
Could the tag be updated to reflect that this issue affects the latest nightly release, so it can be evaluated further? 
@jonkoops @ssilvert 
Please let me know if there’s any additional information or testing you need from my side to assist with this.

Thank you for your continued support!

**Comment by rmartinc** [comment]
I'm moving this to UI because, after testing, the different behavior is at account console level. In keycloak 16 (account v1, so very long ago) the console managed a `kc_locale` parameter that allowed to control of the locale displayed in the console.

For example, when you have logged in the account in keycloak 16 you can change the URL to `http://localhost:8080/auth/realms/test/account/?kc_locale=es` and the account console is in Spanish. @AkshadaMulay seems to have been using this parameter to change the locale. In the current version of the console the locale is not managed at all via parameter but using the attribute in the user. My feeling is that this parameter was never documented and used internally and now it's simply different.

**Comment by AkshadaMulay** [comment]
Thank you for the clarification. I understand that the behavior of the kc_locale parameter has changed in the current account console. Since this was an undocumented feature, can you confirm if there is an alternative way to programmatically set the locale for the account console the latest Keycloak version?
Our use case involves synchronizing the locale between our application and the Keycloak account console, and the previous kc_locale behavior made this seamless. Any guidance on how to achieve this with the current implementation would be greatly appreciated.
Thank you for your support and insights!

**Comment by rmartinc** [comment]
@ssilvert The parameter for the login should be `ui_locales` (for example `ui_locales=es`) and it was the same in keycloak 16. But I think the question is not about the login page but the account console itself. In version 16 adding the `kc_locale` parameter to the account console (after login, when you are inside the app) changed the locale in the app.

@AkshadaMulay I think that now you need to create the user with the locale attribute set. So  the idea would be adding that attribute when you create the user in keycloak.

**Comment by AkshadaMulay** [comment]
Steps We Took to Set the Locale Attribute While Creating a User:
1. Updated the create Method in Users.php

- We modified the create method in our Users.php file to include a locale attribute. Here is the updated code:

    public function create(
        ..Initial key => valuePairs..,
        string $defaultLocale
    ) {
        $uri = sprintf('%s/users', $this->apiRoot());
        try {
            $payload = [
                ..Initial key => valuePairs..,
                'attributes' => [
                    'locale' => $defaultLocale, // Setting the locale attribute
                ],];
            $response = $this->guzzle->request('POST', $uri, [
                'headers' => $this->authHeaders(),
                'json' => $payload,
            ]);
            return true;
        } catch (ClientException $e) {
           ...Exception Handling..
        }
    }

2. Tested User Creation

- We attempted to create a user via the Keycloak Admin API, ensuring that the locale attribute was passed.
- Expected Behavior: The locale attribute should appear in the Keycloak admin dashboard under the user's "Attributes" section.

3. Observed the Outcome

- Despite passing the locale attribute during user creation, it did not appear in the "Attributes" section of the Keycloak dashboard.
- Consequently, the kc_locale parameter had no effect, and the profile page continued to use the default language set in Keycloak.

4. We performed a similar action on Update method of our users class.
5. Tried Updating the Locale Attribute for Existing Users

- To test the update method, we created a new API route to update the locale attribute for a specific user:

    Route::post('/keycloak/update-locale', function (Request $request) {
        $keycloakId = $request->input('keycloakId');
        $locale = fetched request locale here
        $users = fetched respective user here
        $result = $users->update($keycloakId, '', '', '', ['locale' => $locale]);
        if ($result === true) {
            return response()->json(..Success Message..);
        }
        return response()->json(..Failure Message..);
    });

6. Observations

- Despite all these changes, the kc_locale parameter did not work for:
- Newly created users (even with locale set).
- Existing users updated with the locale attribute.
- The profile page continues to use the default language set in Keycloak, ignoring the kc_locale parameter.


7. Questions/Request for Guidance

- Are there additional configurations required in Keycloak to make the locale attribute influence the kc_locale parameter?
- Could this be some kind of a regression issue in Keycloak 25?
- Is there a recommended approach to dynamically sync the kc_locale parameter with the application UI language?

Thanks in advance for your support!

**Comment by rmartinc** [comment]
@AkshadaMulay I don't fully understand you. The idea in the new v3 account console is that the locale used is the one defined in the user (or default if not assigned). You don't need to pass any `kc_locale` parameter anymore.

So for example, if I create a new user like this (using kcadm):

```bash
./kcadm.sh create users -r test -s username=test -s enabled=true -s firstName=test -s lastName=test -s email=test@test.com -s attributes.locale=es
Created new user with id '30f734f8-f8e1-4dfc-9bd4-77185c83232f'
./kcadm.sh set-password --username test --new-password secret -r test
./kcadm.sh get users/30f734f8-f8e1-4dfc-9bd4-77185c83232f -r test
{
  "id" : "30f734f8-f8e1-4dfc-9bd4-77185c83232f",
  "username" : "test",
  "firstName" : "test",
  "lastName" : "test",
  "email" : "test@test.com",
  "emailVerified" : false,
  "attributes" : {
    "locale" : [ "es" ]
  },
  "createdTimestamp" : 1736787223505,
  "enabled" : true,
  "totp" : false,
  "disableableCredentialTypes" : [ ],
  "requiredActions" : [ ],
  "notBefore" : 0,
  "access" : {
    "manageGroupMembership" : true,
    "view" : true,
    "mapRoles" : true,
    "impersonate" : true,
    "manage" : true
  }
}
```

You have to see the  `"locale" : [ "es" ]` here (and in the admin console, it should be the same). If you don't see it, the locale is not assigned OK.

With that if I access the account console: http://localhost:8080/realms/test/account. The login page should be presented in the default language or the one determined by the browser (because the locale is assigned to the user and here user is not known yet). At this point (login to account console) you can modify the locale using the standard parameter `ui_locales` (http://localhost:8080/realms/test/protocol/openid-connect/auth?ui_locales=es&client_id=account-console&redirect_uri=...). But I think this is not what you are asking for.

Once you enter username and password the account console is displayed in the locale assigned to the user, `es` in my case. Just because the locale is assigned OK to the test user.

I'm testing with 25.0.6. I also have checked that (at least in Spanish) the translation in 25 is missing a lot of keys and in 26 is much more complete.

**Comment by mposolda** [comment]
IMO this is not a bug, but rather an enhancement to support this.

- The `kc_locale` is an internal Keycloak parameter, which is not documented anywhere. The fact that there was some behaviour regarding this parameter in Account console V1 does not imply that there should be same behaviour in newest Keycloak 26. It is not documented thing and used just for the internal purposes

- Maybe we can consider something like "forwarding" the `ui_locales` parameter (which is OIDC standard parameter) from the account console URI to the OIDC login URI? So for example when someone opens the http://localhost:8080/realms/test/account?ui_locales=es , then it will be forwarded to the login URL and it will show something like http://localhost:8080/realms/test/protocol/openid-connect/auth?ui_locales=es&client_id=account-console&redirect_uri=... , which means login page will be in the requested locale (es). AFAIK we support something like this for the `scope` parameter for account console as well. However I see this as an enhancement rather than a bug and having a very low priority IMO.

**Comment by jonkoops** [comment]
I agree with @mposolda's assessment, I'll re-label this to an enhancement. Either way, this will be low priority, so the team is unlikely to have to the time to pick this up. We'd welcome a community contribution if this is important to you.

**Comment by rmartinc** [comment]
@jonkoops @mposolda I think this this issue is just about the account console. Not the locale presented at login when accessing the account console. We can do what @mposolda suggest about adding `ui_locales` to the console URL, I think it's find but that would be another issue. @AkshadaMulay  is asking about the locale of the account console itself (after login). Now the account console just shows the locale of the user or default.

**Comment by AkshadaMulay** [comment]
Thank you all for your time and support, especially @rmartinc for suggesting to use attributes.locale to synchronize the profile language with the Keycloak account console. This is a great approach and aligns with Keycloak's current behavior.

However, since there is no "Select Locale" option during user creation on our end, the attributes.locale parameter wasn’t being mapped or passed correctly to Keycloak. As a result, the "Select a locale" dropdown in the Keycloak user details remained empty.

We appreciate the suggestion and will explore implementing it with some modifications to our application, including adding a locale selection feature to ensure better synchronization.
Thank you again for your valuable guidance!

**Comment by mrleblanc101** [comment]
I see this is issue has been closed, was it ever fixed ?
I currently use `ui_locales` to match the Keycloak `login` and `register` page to my app locale.
But it doesn't work when I redirect the user to the `account` page.
I'm still on Keycloak 25, and would like to know if this is fixed, otherwise I have no need to update.

**Comment by ssilvert** [comment]
> I see this is issue has been closed, was it ever fixed ? I currently use `ui_locales` to match the Keycloak `login` and `register` page to my app locale. But it doesn't work when I redirect the user to the `account` page. I'm still on Keycloak 25, and would like to know if this is fixed, otherwise I have no need to update.

@mrleblanc101 I don't think any action was taken for this and there is no PR associated with the issue.

**Comment by mrleblanc101** [comment]
@edewit Can we know why this was closed "as completed" since there has been no resolution ? 
Seems like a basic functionality which is missing from `keycloak.v3` account theme

**Comment by eproulx-cfl** [comment]
Can you please explain why this was closed as "completed" but without any change? @edewit

**Comment by edewit** [comment]
@eproulx-cfl I don't remember why I closed it, it was more then a year ago... I'll re-open it as it wasn't fixed as far as I know

**Comment by Jeeva-z** [comment]
@edewit Kindly refer https://github.com/keycloak/keycloak/discussions/48762, as this might be a duplicate issue

**Comment by Richard31415926** [comment]
Hi, I'd like to work on this issue. I've read the description and would be happy to submit a PR. Is it still available?

**Comment by edewit** [comment]
I'm not sure if this really is an issue, we used to have this kc_locale in the url. But we have a lot of ways to set the locale and it has changed. But I think the current way we use it we don't need this url parameter. I think it works like this now: browser set locale -> user set locale -> default locale. Also @Richard31415926 there is already an PR for this so....

**Comment by scandinave** [comment]
The PR #49278 is closed without merge with no reason.

**Comment by edewit** [comment]
@scandinave it seems to be closed by the creator of said PR @darwvin-dev maybe after reading my comment

**Comment by darwvin-dev** [comment]
@edewit  Thanks for working on this. One small concern: does reading `kc_locale` directly in the UI still handle unsupported/regional locales and the normal locale precedence correctly?

I explored the same issue in #49278 using `LocaleUtil.processLocaleParam()`. No need to reopen it, but its tests might be useful here.

**Comment by scandinave** [comment]
@darwvin-dev I think there is a misunderstanding. Edewit was talking about your PR. So I think you are the only one working on this issue.

**Comment by mrleblanc101** [comment]
@scandinave Mora140 linked a PR 3 days ago

**Comment by scandinave** [comment]
Ohh sorry I missed it

**Comment by edewit** [comment]
The way I see is that `kc_locale` is a internal to save the state of what the user selected on the login page and after a login,  `UpdateUserLocaleAction` saves it to the user profile (locale attribute) and updates the cookie. So why would we need the `kc_locale` support in the account ui? Right now it uses this to find out what language to use:

1. Auth-session locale_user_requested (kc_locale during login)
2. User profile locale attribute
3. Locale cookie
4. Accept-Language header
5. Realm default locale
6. English fallback

**Comment by edewit** [comment]
Before we decide whether to restore kc_locale on the Account Console URL, we would like to better  understand your integration setup. Your use case appears to rely on passing kc_locale in the account URL instead, which means the application knows the user's locale but Keycloak does not (or they may differ). Could you help us understand why locale lives in two places in your setup?

Understanding this will help us judge whether we should add URL-based locale override support to Account v3, or whether the better path is documentation and a supported sync pattern from the application to the user's Keycloak profile.

**Comment by darwvin-dev** [comment]
Makes sense. I’ll leave the integration/use-case question to the reporter.

If URL-based locale override is something we decide to support, my previous approach in #49278 routed `kc_locale` through `LocaleUtil.processLocaleParam()` before `resolveLocale()`, so supported-locale validation and the existing server-side precedence are preserved.

Happy to revive/adapt that approach if that ends up being the desired behavior.
